src/OpenLoco/src/World/Company.cpp

```diff
@@ -373,6 +375,255 @@ namespace OpenLoco
+    }
+
+    struct ProfitAndValue
+    {
+        currency48_t vehicleProfit;
+        currency48_t companyValue;
+    };
+    // 0x00437D79
+    static ProfitAndValue calculateCompanyValue(const Company& company)
+    {
+        if ((company.challengeFlags & CompanyFlags::bankrupt) != CompanyFlags::none)
+        {
+            return { 0, 0 };
+        }
+
+        currency48_t totalValue = company.cash;
+        totalValue -= company.currentLoan;
+
+        currency48_t totalVehicleProfit = 0;
+        for (auto head : VehicleManager::VehicleList())
+        {
+            if (head->owner != company.id())
+            {
+                continue;
+            }
+            if (head->has38Flags(Vehicles::Flags38::isGhost))
+            {
+                continue;
+            }
+
+            Vehicles::Vehicle train(*head);
+            currency32_t trainProfit = train.veh2->totalRecentProfit();
+            // Unsure why /2
+            totalVehicleProfit += trainProfit / 2;
+
+            totalValue += trainProfit * 8;
+
+            for (auto& car : train.cars)
+            {
+                totalValue += car.front->refundCost;
+            }
+        }
+        totalValue = std::max<currency48_t>(0, totalValue);
+        return { totalVehicleProfit, totalValue };
+    }
+
+    // 0x004389CC
+    static void stopAllCompanyVehicles(const CompanyId companyId)
+    {
+        const auto prevUpdateCompany = GameCommands::getUpdatingCompanyId();
+        GameCommands::setUpdatingCompanyId(companyId);
+
+        for (auto head : VehicleManager::VehicleList())
+        {
+            if (head->owner != companyId)
+            {
+                continue;
+            }
+
+            if (head->has38Flags(Vehicles::Flags38::isGhost))
+            {
+                continue;
+            }
+            GameCommands::VehicleChangeRunningModeArgs args{};
+            args.head = head->id;
+            args.mode = GameCommands::VehicleChangeRunningModeArgs::Mode::stopVehicle;
+            // This used to call the command directly
+            GameCommands::doCommand(args, GameCommands::Flags::apply);
+        }
+
+        GameCommands::setUpdatingCompanyId(prevUpdateCompany);
+    }
+
+    void Company::updateMonthly1()
+    {
+        std::rotate(std::begin(cargoUnitsDeliveredHistory), std::end(cargoUnitsDeliveredHistory) - 1, std::end(cargoUnitsDeliveredHistory));
+        cargoUnitsDeliveredHistory[0] = cargoUnitsTotalDelivered;
+        cargoUnitsTotalDelivered = 0;
+
+        std::rotate(std::begin(cargoUnitsDistanceHistory), std::end(cargoUnitsDistanceHistory) - 1, std::end(cargoUnitsDistanceHistory));
+        cargoUnitsDistanceHistory[0] = cargoUnitsTotalDistance;
+        cargoUnitsTotalDistance = 0;
+
+        if (historySize >= 2)
+        {
+            if (cargoUnitsDeliveredHistory[0] < cargoUnitsDeliveredHistory[1]
+                && cargoUnitsDeliveredHistory[1] < cargoUnitsDeliveredHistory[2])
```

Comment:
If `historySize` is 2 then `cargoUnitsDeliveredHistory[2]` would be out of bounds?

Reply:
It's fine because history size is actually 3 at this point we have just shuffled in the additional history but haven't updated history size. I had the same thought.

Reply:
In this case I think we should have a comment explaining that as this question might pop up again.